File: frontend/src/utils/api/cloudAnnouncements.ts
/**
 * Hosted notice-board unread read — cloud mode only.
 *
 * Sibling of `cloudCredits.ts`, and deliberately just as thin: the player reads
 * announcements in the Portal, so all that crosses this boundary is whether
 * something is waiting. No titles, no bodies, no locales.
 *
 * Two failure modes must stay distinguishable, because the dot reacts to them
 * differently:
 *
 *   - **there is no board** — self-host, or a cloud operator with no projected
 *     tenant. The dot is hidden permanently. This is a 404 that *names itself*
 *     (`detail.reason`); a bare 404 does not qualify, because during a rolling
 *     deploy a new SPA routinely reaches an older replica that does not have
 *     this route yet, and treating that as permanent would switch the dot off
 *     for the rest of the session over a transient topology fact.
 *   - **momentarily unknown** — 503, transport failure, an unnamed 404, or a
 *     response we cannot read. Whatever was last known stays, rather than a dot
 *     blinking out and hiding a notice.
 */

import { authedFetch } from '@/utils/authedFetch'

export interface CloudAnnouncementsSnapshot {
  has_unread: boolean
  unread_count: number
  latest_published_at: string | null
  /** True when the backend served last-known-good after an upstream miss. */
  stale: boolean
}

export type CloudAnnouncementsResult =
  | { kind: 'ok'; snapshot: CloudAnnouncementsSnapshot }
  /** No board for this deployment / operator — stop showing the dot. */
  | { kind: 'unsupported' }
  /** Temporarily unknown — keep whatever was last known. */
  | { kind: 'degraded' }

const ENDPOINT = '/api/v1/cloud/announcements/unread'

/** 404 reasons that really do mean "this deployment has no notice board". */
const NO_BOARD_REASONS = new Set(['not_cloud_mode', 'no_cloud_tenant'])

export async function fetchCloudAnnouncements(
  options: { refresh?: boolean } = {},
): Promise<CloudAnnouncementsResult> {
  let res: Response
  try {
    res = await authedFetch(options.refresh ? `${ENDPOINT}?refresh=1` : ENDPOINT)
  } catch {
    return { kind: 'degraded' }
  }
  if (res.status === 404) {
    return (await namesNoBoard(res)) ? { kind: 'unsupported' } : { kind: 'degraded' }
  }
  if (!res.ok) return { kind: 'degraded' }
  let payload: unknown
  try {
    payload = await res.json()
  } catch {
    return { kind: 'degraded' }
  }
  const snapshot = parse(payload)
  return snapshot === null ? { kind: 'degraded' } : { kind: 'ok', snapshot }
}

/** True only for a 404 the endpoint itself produced, naming why there is no board. */
async function namesNoBoard(res: Response): Promise<boolean> {
  try {
    const body = (await res.json()) as { detail?: { reason?: unknown } }
    return (
      typeof body?.detail?.reason === 'string' && NO_BOARD_REASONS.has(body.detail.reason)
    )
  } catch {
    return false
  }
}

/**
 * Strict: a payload missing a field, or carrying the wrong type, or contradicting
 * itself, is `null` — which routes to `degraded`.
 *
 * Coercing junk into a well-formed `has_unread: false` is the single worst thing
 * this file could do. It looks like a successful read, gets stored as the current
 * value, and silently turns off a notice the operator was required to give. There
 * is no symmetric harm in the other direction: a dot that stays up costs a glance.
 */
function parse(payload: unknown): CloudAnnouncementsSnapshot | null {
  if (payload === null || typeof payload !== 'object') return null
  const body = payload as Record<string, unknown>
  const hasUnread = body.has_unread
  const unreadCount = body.unread_count
  if (typeof hasUnread !== 'boolean') return null
  if (
    typeof unreadCount !== 'number'
    || !Number.isInteger(unreadCount)
    || unreadCount < 0
  ) {
    return null
  }
  // The two fields state the same fact; disagreement means this is not one
  // coherent answer, so neither half is trustworthy.
  if (hasUnread !== unreadCount > 0) return null
  const latest = body.latest_published_at
  if (latest !== null && latest !== undefined && typeof latest !== 'string') return null
  return {
    has_unread: hasUnread,
    unread_count: unreadCount,
    latest_published_at: typeof latest === 'string' ? latest : null,
    stale: body.stale === true,
  }
}

File: src/kokoro_link/api/routes/cloud_announcements.py
"""Player-facing notice-board unread proxy — cloud mode only.

Sibling of ``GET /cloud/credits``: same reason for existing (the SPA cannot
reach the User service directly), same cloud-mode-only 404 so self-host's API
inventory is unchanged, same fail-soft contract.

This endpoint carries **one bit and no prose**. Players read announcements in
the Portal; Core only tells them something is waiting and links out. That keeps
the notices in a single place with a single set of translations, and means Core
has nothing to render, localise, or keep in sync.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel

from kokoro_link.api.dependencies import (
    get_container,
    get_current_user,
    is_cloud_mode,
)
from kokoro_link.bootstrap.container import ServiceContainer
from kokoro_link.domain.entities.operator_profile import OperatorProfile

router = APIRouter(prefix="/cloud", tags=["cloud"])


def _no_board(reason: str, message: str) -> HTTPException:
    """A 404 that says *why* there is no board.

    The reason code exists because "this deployment has no notice board" and "this
    replica has not been upgraded yet" arrive at the browser as the same bare 404,
    and the SPA treats the first as permanent. During a rolling deploy a new SPA
    routinely reaches an older replica that lacks this route, and without something
    to tell the two apart the dot would switch itself off for the rest of the
    session over a transient topology fact.
    """
    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail={"reason": reason, "message": message},
    )


class CloudAnnouncementsResponse(BaseModel):
    """Dot payload. Every announcement body stays in the Portal."""

    has_unread: bool
    unread_count: int
    latest_published_at: str | None = None
    stale: bool
    """True when the upstream read failed and this is last-known-good."""


@router.get("/announcements/unread", response_model=CloudAnnouncementsResponse)
async def get_cloud_announcements_unread(
    refresh: bool = Query(
        default=False,
        description=(
            "Bypass the cache. The SPA sets this when the player returns to the "
            "tab, where they may have just read the board in the Portal."
        ),
    ),
    user: OperatorProfile = Depends(get_current_user),
    container: ServiceContainer = Depends(get_container),
) -> CloudAnnouncementsResponse:
    if not is_cloud_mode(container):
        raise _no_board("not_cloud_mode", "announcements are only available in cloud mode")
    tenant_id = (getattr(user, "cloud_tenant_id", None) or "").strip()
    if not tenant_id:
        # A cloud-mode operator with no projected tenant has no read state at
        # all — no board, so the dot stays hidden.
        raise _no_board("no_cloud_tenant", "operator has no cloud tenant")
    service = getattr(container, "cloud_announcement_service", None)
    snapshot = (
        await service.get_unread(tenant_id, refresh=refresh)
        if service is not None
        else None
    )
    if snapshot is None:
        # Nothing known and nothing cached. 503 rather than "nothing unread":
        # inventing a clear board would hide a notice the operator is required
        # to have given.
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="announcements are temporarily unavailable",
        )
    return CloudAnnouncementsResponse(
        has_unread=snapshot.has_unread,
        unread_count=snapshot.unread_count,
        latest_published_at=snapshot.latest_published_at,
        stale=snapshot.stale,
    )
